### ISSUE TITLE
More abstract classes for a broader class of tasks

### ISSUE THREAD

**Issue by saleml** [saleml/torchgfn#18]
The current design of the library is too specific to environments with discrete actions that can be masked out in a straightforward way.

Following a discussion with @EricElmoznino, the library would be much more flexible if there is a more abstract set of classes and methods. While this might require significant changes in how the states and environments and linked, and how that is used for sampling trajectories and evaluating losses, the core components of the library (sampling, loss evaluation) are already written.

Suggestions made by @EricElmoznino:
----
The current `ActionsSampler`'s are too restrictive: for instance, in environments with continuous action spaces, the action could correspond to a vector of floats, and the distribution over actions might be something like a tuple of the mean and standard deviation over actions predicted by the forward policy. Another example is when your states are graphs and you want to perform some kind of graph surgery. Your actions might be something like "remove the node at a particular location" or "join a set of nodes into a clique", where you would want more complex data structures to represent these actions and the distributions over them. Hence:

- [ ] We should have a more abstract `Action` class, where : 
- The `Env.step()` method could then take `Action` instance as argument.
- The `ActionSampler.sample()` method could output `(log_prob, action)` tuples, where `log_prob` is the log probability of the sampled action. This would be very similar to the current behaviour, except that the actions are no longer constrained to be integer tensors (they would be instances of the `Action` class).
- The `FunctionEstimator.__call__()` method could output arbitrary datatype(s) rather than just tensors. If we wanted to be very verbose, we could also have the forward and backward policy subclasses output something like an ActionDistribution subclass, which different ActionSampler subclasses would take as input in their .sample() methods.
For instance, we could have a `CategoricalActionDistribution` class. A `DiscreteLogitPFEstimator` subclass that outputs a `CategoricalActionDistribution`, and a `DiscreteActionSampler` that requires an instance of `DiscreteLogitPFEstimator `at initialization. As another example, we could have a `GaussianActionDistribution` class (essentially consisting of mean and variance), a `GaussianLogitPFEstimator` subclass that outputs a `GaussianActionDistribution`, and a `GaussianActionSampler` that requires an instance of `GaussianLogitPFEstimator` at initialization.
You could imagine extending this for your particular project that uses an idiosyncratic action space and distribution.
- For now, masks of the type `actions != -1` or `actions == env.n_actions - 1` are used in multiple places (notably during trajectory sampling). These would obviously need to be changed to something like `actions.is_dummy()` and `actions.is_exit()` where the abstract `Actions` class somehow defines what’s the dummy action (that’s appended to short trajectories), and what’s the exit action (i.e. actions corresponding to s -> s_f transitions)

- [ ] Having a more abstract `PFEstimator` class with an abstract `get_actions_distribution`, which should be used for `ActionsSampler`s and `TrajectoriesSampler`s. The current estimators can be subclasses of `PFEstimator` that would work well for simple discrete environment like HyperGrid, but the user would have the ability to write their own `PFEstimator` that has access to the user-defined environment.

Currently, `FunctionEstimator.__call__(states: States) -> OutputTensor` passes the states through an instance of `Preprocessor` in order to transform the abstract state into something that can be processed by a `GFNModule` (often a neural net). However, the output of `Preprocessor` and the input to `GFNModule` must be a single, fixed-sized tensor.
This is too restrictive. In general, a model might take in more complex data structures. For instance, if the state represents a graph, the `GFNModule` might want to take in a tuple of `(node_attributes: Tensor, adjacency_matrix: Tensor)`. The used should be able to use arguments that are not tensors. This would also be more consistent with how generic nn.Module's work, where they can take in arbitrary arguments. Hence:

- [ ] The `Preprocessor` should output a tuple of arbitrary datatypes and then we pass `*the_tuple` to the `GFNModule`.

----

Going forward, we could have a separate branch on which such features are implemented and tested (on the current environments at least), and this issue update as we discover what other significant changes would be needed.

**Comment by saleml** [comment]
All of this should be handled by this PR https://github.com/saleml/torchgfn/pull/24
